tasks/insights-client-unregistration.yml

```diff
@@ -0,0 +1,22 @@
+# SPDX-License-Identifier: MIT
+---
+- name: Check if insights-client is installed
```

Comment:
Where do you use the package facts?

Reply:
the task after this one runs `insights-client --status`, and that is done only if `insights-client` is installed; maybe we need to add the "package" fact to `__rhc_required_facts`?

Reply:
Unless there is some reason to only conditionally install the insights-clients package, I would just use a `package:` task here:
```yaml
- name: Ensure insights-client is installed
  package:
    name: insights-client
    state: present
```

Reply:
I guess the idea is to not install install `insights-client` if `rhc_insights.status: absent`. Also, one important thing to note here is that `insights-client` is not installed by default in RHEL (it's in AppStream), while `subscription-manager` is in BaseOS and even part of the `@core` package group (so installed in every RHEL installation); because of this, can be installed `insights-client` only _after_ the system is registered. Hence, I think the tasks before `subscription-manager` runs have to cope with missing packages.

Reply:
> I guess the idea is to not install install `insights-client` if `rhc_insights.status: absent`.

Ok.

> Also, one important thing to note here is that `insights-client` is not installed by default in RHEL (it's in AppStream), while `subscription-manager` is in BaseOS and even part of the `@core` package group (so installed in every RHEL installation); because of this, `insights-client` can be installed only _after_ the system is registered. Hence, I think the tasks before `subscription-manager` runs have to cope with missing packages.

Why do the tasks in `insights-client-unregistration.yml` have to come before `subscription-manager.yml`?

Reply:
Is it because if you run `insights-client --unregister` after unsubscribing the system, the command will fail?  If so, is there an error message that can be interpreted to mean "system unsubscribed from rhc" that is different than "system is unregistered with insights" vs. some other error messages?

It seems to me overly complicated to have both `insights-client-unregistration.yml` and `insights-client.yml` with code that is mostly the same.

Reply:
> Is it because if you run `insights-client --unregister` after unsubscribing the system, the command will fail?

Yes, exactly. Cleanly unregister from Insights before the rhsm unregistration is better than unregistering from rhsm and handling the connection failure of `insights-client` (that cannot authenticate anymore) -- it should properly remove the system from the Insights inventory. The joy of having two tools that need to do closely related tasks and are not aware of each other's presence, you see...

Reply:
I agree with all @ptoscano has said and also add that the task needs to check if if the package is install to run the commands, otherwise it will fail. 

Additionally, unregistration of `insights-client` needs to be done before unregister `susbscription-manager` it would not fail but is not a clean unregistration. But it needs that `subscription-manager` is registered to register `insights-client`. That is is why it split the logic.